File: src/protocols/java/TemporalReasoningProtocol.java
package com.medinatech.mmims.protocols;

import java.util.*;
import java.util.stream.Collectors;

/**
 * PROTO-232 — Temporal Reasoning Protocol (Java)
 * Phi-scaled time perception, causal inference, and temporal abstraction
 * for MMIMS-X.
 *
 * Charter: PROTO-232
 * Attribution: Alfredo Medina Hernandez | Medina Tech | Dallas TX | May 2026
 */
public class TemporalReasoningProtocol {

    // ── Constants ──────────────────────────────────────────────────────────
    public static final double PHI     = (1 + Math.sqrt(5)) / 2.0;
    public static final double PHI_INV = 1.0 / PHI;

    // ── 8 Phi-Scaled Time Scales ───────────────────────────────────────────
    public static class TimeScale {
        public final int    index;
        public final String name;
        public final double ms;
        public final String label;
        public TimeScale(int i, String name, double ms, String label) {
            this.index = i; this.name = name; this.ms = ms; this.label = label;
        }
    }

    public static final TimeScale[] TIME_SCALES;
    static {
        String[] labels = {"flash","micro","pulse","breath","beat","wave","cycle","epoch"};
        TIME_SCALES = new TimeScale[8];
        for (int k = 0; k < 8; k++) {
            double ms = 100.0 * Math.pow(PHI, k);
            TIME_SCALES[k] = new TimeScale(k, "τ" + k, ms, labels[k]);
        }
    }

    // ── Temporal Event ─────────────────────────────────────────────────────
    public static class TemporalEvent {
        public final String       id;
        public final String       type;
        public final Object       payload;
        public final long         timestamp;
        public final List<String> causes  = new ArrayList<>();
        public final List<String> effects = new ArrayList<>();
        public       double       weight  = 1.0;
        public       int          scaleIdx = 0;

        public TemporalEvent(String id, String type, Object payload) {
            this.id = id; this.type = type; this.payload = payload;
            this.timestamp = System.currentTimeMillis();
        }

        public void addCause(String eid)  { if (!causes.contains(eid))  causes.add(eid); }
        public void addEffect(String eid) { if (!effects.contains(eid)) effects.add(eid); }

        public double decayedWeight(long nowMs, double halfLifeMs) {
            double age    = nowMs - timestamp;
            double lambda = Math.log(2) / halfLifeMs * PHI_INV;
            return weight * Math.exp(-lambda * age);
        }

        public Map<String, Object> toSummary() {
            Map<String, Object> m = new LinkedHashMap<>();
            m.put("id",        id);
            m.put("type",      type);
            m.put("scale",     TIME_SCALES[Math.min(scaleIdx, 7)].label);
            m.put("causes",    causes.size());
            m.put("effects",   effects.size());
            m.put("timestamp", timestamp);
            return m;
        }
    }

    // ── Phi-Decay Buffer ───────────────────────────────────────────────────
    public static class PhiDecayBuffer {
        private final int                  capacity;
        private final double               halfLifeMs;
        private final Map<String, TemporalEvent> events = new LinkedHashMap<>();

        public PhiDecayBuffer(int capacity, double halfLifeMs) {
            this.capacity   = capacity;
            this.halfLifeMs = halfLifeMs;
        }

        public PhiDecayBuffer() { this(144, TIME_SCALES[4].ms); }

        public void add(TemporalEvent event) {
            if (events.containsKey(event.id)) return;
            if (events.size() >= capacity) evict();
            events.put(event.id, event);
        }

        private void evict() {
            long nowMs = System.currentTimeMillis();
            String minId = null; double minW = Double.MAX_VALUE;
            for (Map.Entry<String, TemporalEvent> e : events.entrySet()) {
                double w = e.getValue().decayedWeight(nowMs, halfLifeMs);
                if (w < minW) { minW = w; minId = e.getKey(); }
            }
            if (minId != null) events.remove(minId);
        }

        public TemporalEvent get(String id)  { return events.get(id); }
        public int size()                    { return events.size(); }
        public Collection<TemporalEvent> all() { return events.values(); }

        public int prune(double threshold) {
            long nowMs  = System.currentTimeMillis();
            int  pruned = 0;
            Iterator<Map.Entry<String, TemporalEvent>> it = events.entrySet().iterator();
            while (it.hasNext()) {
                Map.Entry<String, TemporalEvent> e = it.next();
                if (e.getValue().decayedWeight(nowMs, halfLifeMs) < threshold) {
                    it.remove(); pruned++;
                }
            }
            return pruned;
        }
    }

    // ── Causal Graph ───────────────────────────────────────────────────────
    public static class CausalGraph {
        private final Map<String, Map<String, Double>> outEdges = new HashMap<>();
        private final Map<String, Map<String, Double>> inEdges  = new HashMap<>();

        private void ensureNode(String id) {
            outEdges.putIfAbsent(id, new HashMap<>());
            inEdges.putIfAbsent(id, new HashMap<>());
        }

        public void link(String cause, String effect, double strength) {
            ensureNode(cause); ensureNode(effect);
            outEdges.get(cause).put(effect, strength);
            inEdges.get(effect).put(cause, strength);
        }

        public List<Map<String,Object>> predictEffects(String causeId, int maxDepth) {
            Map<String, Double> visited = new HashMap<>();
            Queue<Object[]> queue = new LinkedList<>();
            queue.add(new Object[]{ causeId, 1.0, 0 });
            while (!queue.isEmpty()) {
                Object[] cur   = queue.poll();
                String   id    = (String)  cur[0];
                double   str   = (double)  cur[1];
                int      depth = (int)     cur[2];
                if (depth >= maxDepth) continue;
                Map<String, Double> out = outEdges.get(id);
                if (out == null) continue;
                for (Map.Entry<String, Double> e : out.entrySet()) {
                    double cumStr = str * e.getValue() * PHI_INV;
                    if (!visited.containsKey(e.getKey()) || visited.get(e.getKey()) < cumStr) {
                        visited.put(e.getKey(), cumStr);
                        queue.add(new Object[]{ e.getKey(), cumStr, depth + 1 });
                    }
                }
            }
            visited.remove(causeId);
            return visited.entrySet().stream()
                .map(e -> { Map<String,Object> m = new LinkedHashMap<>(); m.put("id", e.getKey()); m.put("strength", e.getValue()); return m; })
                .sorted((a,b) -> Double.compare((double)b.get("strength"), (double)a.get("strength")))
                .collect(Collectors.toList());
        }

        public List<Map<String,Object>> inferCauses(String effectId, int maxDepth) {
            Map<String, Double> visited = new HashMap<>();
            Queue<Object[]> queue = new LinkedList<>();
            queue.add(new Object[]{ effectId, 1.0, 0 });
            while (!queue.isEmpty()) {
                Object[] cur   = queue.poll();
                String   id    = (String) cur[0];
                double   str   = (double) cur[1];
                int      depth = (int)    cur[2];
                if (depth >= maxDepth) continue;
                Map<String, Double> in = inEdges.get(id);
                if (in == null) continue;
                for (Map.Entry<String, Double> e : in.entrySet()) {
                    double cumStr = str * e.getValue() * PHI_INV;
                    if (!visited.containsKey(e.getKey()) || visited.get(e.getKey()) < cumStr) {
                        visited.put(e.getKey(), cumStr);
                        queue.add(new Object[]{ e.getKey(), cumStr, depth + 1 });
                    }
                }
            }
            visited.remove(effectId);
            return visited.entrySet().stream()
                .map(e -> { Map<String,Object> m = new LinkedHashMap<>(); m.put("id", e.getKey()); m.put("strength", e.getValue()); return m; })
                .sorted((a,b) -> Double.compare((double)b.get("strength"), (double)a.get("strength")))
                .collect(Collectors.toList());
        }

        public int nodeCount() { return outEdges.size(); }
        public int edgeCount() { return outEdges.values().stream().mapToInt(Map::size).sum(); }
    }

    // ── Temporal Abstractor ────────────────────────────────────────────────
    public static class TemporalAbstractor {
        public List<Map<String, Object>> summarise(List<TemporalEvent> events, int scaleIndex) {
            if (events.isEmpty()) return Collections.emptyList();
            double windowMs = TIME_SCALES[Math.min(scaleIndex, 7)].ms;
            List<TemporalEvent> sorted = events.stream()
                .sorted(Comparator.comparingLong(e -> e.timestamp))
                .collect(Collectors.toList());

            List<Map<String, Object>> episodes = new ArrayList<>();
            List<TemporalEvent> bucket = new ArrayList<>();
            long bucketStart = sorted.get(0).timestamp;

            for (TemporalEvent ev : sorted) {
                if (ev.timestamp - bucketStart <= windowMs) {
                    bucket.add(ev);
                } else {
                    episodes.add(mergeEpisode(bucket, scaleIndex));
                    bucket = new ArrayList<>(List.of(ev));
                    bucketStart = ev.timestamp;
                }
            }
            if (!bucket.isEmpty()) episodes.add(mergeEpisode(bucket, scaleIndex));
            return episodes;
        }

        private Map<String, Object> mergeEpisode(List<TemporalEvent> evs, int scaleIndex) {
            Set<String> types   = evs.stream().map(e -> e.type).collect(Collectors.toSet());
            Set<String> causes  = evs.stream().flatMap(e -> e.causes.stream()).collect(Collectors.toSet());
            Set<String> effects = evs.stream().flatMap(e -> e.effects.stream()).collect(Collectors.toSet());
            long startMs = evs.get(0).timestamp;
            long endMs   = evs.get(evs.size() - 1).timestamp;
            Map<String, Object> m = new LinkedHashMap<>();
            m.put("scaleLabel",  TIME_SCALES[Math.min(scaleIndex, 7)].label);
            m.put("eventCount",  evs.size());
            m.put("types",       new ArrayList<>(types));
            m.put("startMs",     startMs);
            m.put("endMs",       endMs);
            m.put("durationMs",  endMs - startMs);
            m.put("causes",      new ArrayList<>(causes));
            m.put("effects",     new ArrayList<>(effects));
            m.put("phiWeight",   evs.size() / TIME_SCALES[Math.min(scaleIndex, 7)].ms * PHI);
            return m;
        }
    }

    // ── Temporal Reasoning Engine ──────────────────────────────────────────
    public static class TemporalReasoningEngine {
        private final PhiDecayBuffer    buffer;
        private final CausalGraph       causal;
        private final TemporalAbstractor abstractor;
        private       int               eventCounter = 0;

        public TemporalReasoningEngine()              { this(144); }
        public TemporalReasoningEngine(int capacity)  {
            this.buffer     = new PhiDecayBuffer(capacity, TIME_SCALES[4].ms);
            this.causal     = new CausalGraph();
            this.abstractor = new TemporalAbstractor();
        }

        public TemporalEvent ingest(String type, Object payload, List<String> causeIds) {
            String        id = "EVT-" + (++eventCounter) + "-" + System.currentTimeMillis();
            TemporalEvent ev = new TemporalEvent(id, type, payload);
            for (String cid : causeIds) {
                ev.addCause(cid);
                causal.link(cid, id, PHI_INV);
                TemporalEvent cause = buffer.get(cid);
                if (cause != null) cause.addEffect(id);
            }
            buffer.add(ev);
            return ev;
        }

        public void assertCausation(String causeId, String effectId, double strength) {
            causal.link(causeId, effectId, strength);
            TemporalEvent c = buffer.get(causeId), e = buffer.get(effectId);
            if (c != null) c.addEffect(effectId);
            if (e != null) e.addCause(causeId);
        }

        public List<Map<String,Object>> predictEffects(String id)    { return causal.predictEffects(id, 3); }
        public List<Map<String,Object>> inferCauses(String id)       { return causal.inferCauses(id, 3); }
        public List<Map<String,Object>> summarise(int scaleIndex)    {
            return abstractor.summarise(new ArrayList<>(buffer.all()), scaleIndex);
        }
        public int prune()    { return buffer.prune(0.01); }

        public Map<String,Object> status() {
            Map<String,Object> m = new LinkedHashMap<>();
            m.put("buffered",     buffer.size());
            m.put("causalNodes",  causal.nodeCount());
            m.put("causalEdges",  causal.edgeCount());
            return m;
        }
    }

    // ── Demo ───────────────────────────────────────────────────────────────
    public static void main(String[] args) throws InterruptedException {
        System.out.println("=== PROTO-232 Temporal Reasoning Protocol (Java) ===");
        System.out.println("Time scales:");
        for (TimeScale s : TIME_SCALES)
            System.out.printf("  %s: %.1f ms (%s)%n", s.name, s.ms, s.label);

        TemporalReasoningEngine engine = new TemporalReasoningEngine();
        TemporalEvent e1 = engine.ingest("sensor-read",  "temperature=98.6", List.of());
        Thread.sleep(10);
        TemporalEvent e2 = engine.ingest("alert-trigger","threshold-exceeded", List.of(e1.id));
        Thread.sleep(5);
        TemporalEvent e3 = engine.ingest("agent-spawn",  "risk-agent-007",  List.of(e2.id));

        System.out.println("\nPredicted effects of " + e1.id + ": " + engine.predictEffects(e1.id));
        System.out.println("Inferred causes of  " + e3.id + ": " + engine.inferCauses(e3.id));
        System.out.println("Summary (beat scale): " + engine.summarise(4));
        System.out.println("Status: " + engine.status());
    }
}

File: src/protocols/python/temporal_reasoning_protocol.py
"""
PROTO-232 — Temporal Reasoning Protocol (Python)
Phi-scaled time perception, causal inference, and temporal abstraction
for MMIMS-X.

Charter: PROTO-232
Attribution: Alfredo Medina Hernandez | Medina Tech | Dallas TX | May 2026
"""

from __future__ import annotations
import math
import time
import uuid
from collections import deque
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Set, Tuple

# ── Constants ──────────────────────────────────────────────────────────────────
PHI     = (1 + math.sqrt(5)) / 2
PHI_INV = 1 / PHI

# ── 8 Phi-Scaled Time Scales ───────────────────────────────────────────────────
_LABELS = ["flash", "micro", "pulse", "breath", "beat", "wave", "cycle", "epoch"]

@dataclass(frozen=True)
class TimeScale:
    index: int
    name:  str
    ms:    float
    label: str

TIME_SCALES: List[TimeScale] = [
    TimeScale(k, f"τ{k}", 100.0 * PHI ** k, _LABELS[k])
    for k in range(8)
]


# ── Temporal Event ─────────────────────────────────────────────────────────────
class TemporalEvent:
    def __init__(self, id: str, type: str, payload: Any,
                 timestamp_ms: Optional[float] = None):
        self.id           = id
        self.type         = type
        self.payload      = payload
        self.timestamp    = timestamp_ms if timestamp_ms is not None else time.time() * 1000
        self.causes:  List[str] = []
        self.effects: List[str] = []
        self.weight       = 1.0
        self.scale_idx    = 0

    def add_cause(self, event_id: str) -> None:
        if event_id not in self.causes:
            self.causes.append(event_id)

    def add_effect(self, event_id: str) -> None:
        if event_id not in self.effects:
            self.effects.append(event_id)

    def decayed_weight(self, now_ms: Optional[float] = None,
                       half_life_ms: Optional[float] = None) -> float:
        if now_ms is None:
            now_ms = time.time() * 1000
        if half_life_ms is None:
            half_life_ms = TIME_SCALES[4].ms
        age    = now_ms - self.timestamp
        lam    = math.log(2) / half_life_ms * PHI_INV
        return self.weight * math.exp(-lam * age)

    def to_summary(self) -> Dict[str, Any]:
        return {
            "id":        self.id,
            "type":      self.type,
            "scale":     TIME_SCALES[min(self.scale_idx, 7)].label,
            "causes":    len(self.causes),
            "effects":   len(self.effects),
            "timestamp": self.timestamp,
        }


# ── Phi-Decay Buffer ───────────────────────────────────────────────────────────
class PhiDecayBuffer:
    def __init__(self, capacity: int = 144,
                 half_life_ms: Optional[float] = None):
        self.capacity    = capacity  # 144 = Fibonacci F₁₂
        self.half_life   = half_life_ms or TIME_SCALES[4].ms
        self._events:    Dict[str, TemporalEvent] = {}
        self._order:     List[str] = []

    def add(self, event: TemporalEvent) -> None:
        if event.id in self._events:
            self._refresh(event.id)
            return
        if len(self._events) >= self.capacity:
            self._evict()
        self._events[event.id] = event
        self._order.append(event.id)

    def _refresh(self, id: str) -> None:
        if id in self._order:
            self._order.remove(id)
            self._order.append(id)

    def _evict(self) -> None:
        now  = time.time() * 1000
        min_id, min_w = None, float("inf")
        for id, ev in self._events.items():
            w = ev.decayed_weight(now, self.half_life)
            if w < min_w:
                min_w, min_id = w, id
        if min_id:
            del self._events[min_id]
            self._order = [i for i in self._order if i != min_id]

    def get(self, id: str) -> Optional[TemporalEvent]:
        return self._events.get(id)

    def size(self) -> int:
        return len(self._events)

    def all(self) -> List[TemporalEvent]:
        return list(self._events.values())

    def ranked(self, now_ms: Optional[float] = None) -> List[Tuple[TemporalEvent, float]]:
        if now_ms is None:
            now_ms = time.time() * 1000
        pairs = [(ev, ev.decayed_weight(now_ms, self.half_life)) for ev in self._events.values()]
        return sorted(pairs, key=lambda x: x[1], reverse=True)

    def prune(self, threshold: float = 0.01) -> int:
        now    = time.time() * 1000
        to_del = [id for id, ev in self._events.items()
                  if ev.decayed_weight(now, self.half_life) < threshold]
        for id in to_del:
            del self._events[id]
        self._order = [i for i in self._order if i in self._events]
        return len(to_del)


# ── Causal Graph ───────────────────────────────────────────────────────────────
class CausalGraph:
    def __init__(self):
        self._out: Dict[str, Dict[str, float]] = {}
        self._in:  Dict[str, Dict[str, float]] = {}

    def _ensure(self, id: str) -> None:
        self._out.setdefault(id, {})
        self._in.setdefault(id, {})

    def link(self, cause: str, effect: str, strength: float = None) -> None:
        if strength is None:
            strength = PHI_INV
        self._ensure(cause)
        self._ensure(effect)
        self._out[cause][effect] = strength
        self._in[effect][cause]  = strength

    def predict_effects(self, cause_id: str, max_depth: int = 3) -> List[Dict[str, Any]]:
        visited: Dict[str, float] = {}
        queue   = [(cause_id, 1.0, 0)]
        while queue:
            id, strength, depth = queue.pop(0)
            if depth >= max_depth:
                continue
            for eff, edge_str in self._out.get(id, {}).items():
                cum = strength * edge_str * PHI_INV
                if eff not in visited or visited[eff] < cum:
                    visited[eff] = cum
                    queue.append((eff, cum, depth + 1))
        visited.pop(cause_id, None)
        return sorted(
            [{"id": k, "strength": v} for k, v in visited.items()],
            key=lambda x: x["strength"], reverse=True,
        )

    def infer_causes(self, effect_id: str, max_depth: int = 3) -> List[Dict[str, Any]]:
        visited: Dict[str, float] = {}
        queue   = [(effect_id, 1.0, 0)]
        while queue:
            id, strength, depth = queue.pop(0)
            if depth >= max_depth:
                continue
            for cause, edge_str in self._in.get(id, {}).items():
                cum = strength * edge_str * PHI_INV
                if cause not in visited or visited[cause] < cum:
                    visited[cause] = cum
                    queue.append((cause, cum, depth + 1))
        visited.pop(effect_id, None)
        return sorted(
            [{"id": k, "strength": v} for k, v in visited.items()],
            key=lambda x: x["strength"], reverse=True,
        )

    def node_count(self) -> int: return len(self._out)
    def edge_count(self) -> int: return sum(len(v) for v in self._out.values())


# ── Temporal Abstractor ────────────────────────────────────────────────────────
class TemporalAbstractor:
    def summarise(self, events: List[TemporalEvent],
                  scale_index: int = 4) -> List[Dict[str, Any]]:
        if not events:
            return []
        ts    = TIME_SCALES[min(scale_index, 7)]
        window = ts.ms
        sorted_evs = sorted(events, key=lambda e: e.timestamp)
        episodes, bucket, start = [], [], sorted_evs[0].timestamp

        for ev in sorted_evs:
            if ev.timestamp - start <= window:
                bucket.append(ev)
            else:
                episodes.append(self._merge(bucket, ts))
                bucket, start = [ev], ev.timestamp
        if bucket:
            episodes.append(self._merge(bucket, ts))
        return episodes

    @staticmethod
    def _merge(evs: List[TemporalEvent], ts: TimeScale) -> Dict[str, Any]:
        types   = list(set(e.type    for e in evs))
        causes  = list(set(c         for e in evs for c in e.causes))
        effects = list(set(x         for e in evs for x in e.effects))
        return {
            "scale_label":  ts.label,
            "event_count":  len(evs),
            "types":        types,
            "start_ms":     evs[0].timestamp,
            "end_ms":       evs[-1].timestamp,
            "duration_ms":  evs[-1].timestamp - evs[0].timestamp,
            "causes":       causes,
            "effects":      effects,
            "phi_weight":   len(evs) / ts.ms * PHI,
        }


# ── Temporal Reasoning Engine ──────────────────────────────────────────────────
class TemporalReasoningEngine:
    def __init__(self, buffer_capacity: int = 144):
        self.buffer     = PhiDecayBuffer(buffer_capacity)
        self.causal     = CausalGraph()
        self.abstractor = TemporalAbstractor()
        self._counter   = 0

    def ingest(self, type: str, payload: Any,
               cause_ids: Optional[List[str]] = None) -> TemporalEvent:
        cause_ids = cause_ids or []
        self._counter += 1
        id = f"EVT-{self._counter}-{int(time.time() * 1000)}"
        ev = TemporalEvent(id, type, payload)
        for cid in cause_ids:
            ev.add_cause(cid)
            self.causal.link(cid, id, PHI_INV)
            cause = self.buffer.get(cid)
            if cause:
                cause.add_effect(id)
        self.buffer.add(ev)
        self.causal._ensure(id)
        return ev

    def assert_causation(self, cause_id: str, effect_id: str,
                         strength: float = None) -> None:
        self.causal.link(cause_id, effect_id, strength or PHI_INV)
        c, e = self.buffer.get(cause_id), self.buffer.get(effect_id)
        if c: c.add_effect(effect_id)
        if e: e.add_cause(cause_id)

    def predict_effects(self, id: str, max_depth: int = 3) -> List[Dict]:
        return self.causal.predict_effects(id, max_depth)

    def infer_causes(self, id: str, max_depth: int = 3) -> List[Dict]:
        return self.causal.infer_causes(id, max_depth)

    def summarise(self, scale_index: int = 4) -> List[Dict]:
        return self.abstractor.summarise(self.buffer.all(), scale_index)

    def prune(self, threshold: float = 0.01) -> int:
        return self.buffer.prune(threshold)

    def status(self) -> Dict[str, Any]:
        return {
            "buffered":      self.buffer.size(),
            "causal_nodes":  self.causal.node_count(),
            "causal_edges":  self.causal.edge_count(),
        }


# ── Quick demo ─────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    print("=== PROTO-232 Temporal Reasoning Protocol (Python) ===")
    for s in TIME_SCALES:
        print(f"  {s.name}: {s.ms:.1f} ms ({s.label})")

    engine = TemporalReasoningEngine()
    e1 = engine.ingest("sensor-read",   "temperature=98.6")
    time.sleep(0.01)
    e2 = engine.ingest("alert-trigger", "threshold-exceeded", [e1.id])
    time.sleep(0.005)
    e3 = engine.ingest("agent-spawn",   "risk-agent-007",    [e2.id])

    print(f"\nPredicted effects of {e1.id}: {engine.predict_effects(e1.id)}")
    print(f"Inferred causes of  {e3.id}: {engine.infer_causes(e3.id)}")
    print(f"Status: {engine.status()}")
